src/Workspaces/Core/Portable/Workspace/Solution/SourceInfoProvider.cs

```diff
@@ -0,0 +1,58 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.CodeAnalysis
+{
+    internal interface ISourceFileInfoProvider
+    {
+        /// <summary>
+        /// return SourceFileInfo for the given full path, if this provide doesn't handle the given file
+        /// return null.
+        /// </summary>
+        Task<SourceFileInfo> GetSourceFileInfoAsync(string fullPath, CancellationToken cancellationToken);
+    }
+
+    /// <summary>
+    /// SourceFileInfo is random name I choose. we might want to chagne this to something like temporarySourceInfo,
+    /// InterimSourceInfo, GeneratedSourceInfo and etc.
+    /// 
+    /// none of info returned from this can be dynamic. it should always return same data
+    /// </summary>
+    internal abstract class SourceFileInfo
+    {
+        /// <summary>
+        /// unique id to identify this file among other files in a project. it should be unique in the project
+        /// </summary>
+        public abstract string Id { get; }
```

Comment:
Are we supposed to generate this? I think we would end up doing like `Guid.NewGuid()` or something silly like that

Reply:
I can remove this. thought might be useful when debugging. but can be removed.